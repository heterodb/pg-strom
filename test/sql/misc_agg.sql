--#
--#       Gpu PreAggregate Complex TestCases. 
--#

set pg_strom.debug_force_gpupreagg to on;
set client_min_messages to warning;
set extra_float_digits to -3;

-- Device Filter
select key, avg(float_x)  from strom_test  where float_x >= 0.5 and float_x <= 0.6 group by key order by key;

-- select distinct
select key,count(distinct smlint_x)  from strom_test group by key order by key;

-- (group by x+y)
select key+1000,count(float_x)  from strom_test group by key+1000 order by key+1000;

-- (group by x::varchar)
select key::varchar(20),count(float_x)  from strom_test group by key::varchar(20) order by key;

-- (group by x::numeric)
select key::numeric,count(float_x)  from strom_test group by key::numeric order by key;

-- aggfunc( x + y + z )
select key, avg(float_x + float_y + float_z)  from strom_mix  group by key order by key;
