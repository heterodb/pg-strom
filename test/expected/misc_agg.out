--#
--#       Gpu PreAggregate Complex TestCases. 
--#
set pg_strom.debug_force_gpupreagg to on;
set client_min_messages to warning;
set extra_float_digits to -3;
-- Device Filter
select key, avg(float_x)  from strom_test  where float_x >= 0.5 and float_x <= 0.6 group by key order by key;
 key |      avg       
-----+----------------
   1 | 0.551919001672
   2 | 0.545772242046
   3 | 0.554197856812
   4 | 0.550778102591
   5 | 0.552478456622
   6 |  0.55536691782
   7 | 0.554688514064
   8 | 0.551600200743
   9 | 0.552619698081
  10 | 0.549357383213
  21 |  0.55716576675
  22 | 0.542070096478
  23 | 0.557485331492
  24 | 0.557206478455
  25 | 0.551280601819
  26 | 0.542253901394
  27 | 0.549630367573
  28 | 0.550002355899
  29 | 0.552391847319
  30 | 0.551167013235
(20 rows)

-- select distinct
select key,count(distinct smlint_x)  from strom_test group by key order by key;
 key | count 
-----+-------
   1 |    34
   2 |    34
   3 |    34
   4 |    34
   5 |    34
   6 |    34
   7 |    34
   8 |    34
   9 |    34
  10 |    34
  11 |    34
  12 |    34
  13 |    34
  14 |    34
  15 |    34
  16 |    34
  17 |    34
  18 |    34
  19 |    34
  20 |    34
  21 |    67
  22 |    67
  23 |    67
  24 |    67
  25 |    67
  26 |    67
  27 |    67
  28 |    67
  29 |    67
  30 |    67
     |     0
(31 rows)

-- (group by x+y)
select key+1000,count(float_x)  from strom_test group by key+1000 order by key+1000;
 ?column? | count 
----------+-------
     1001 |   962
     1002 |   952
     1003 |   957
     1004 |   938
     1005 |   945
     1006 |   948
     1007 |   947
     1008 |   948
     1009 |   954
     1010 |   960
     1011 |   943
     1012 |   952
     1013 |   956
     1014 |   945
     1015 |   956
     1016 |   952
     1017 |   951
     1018 |   955
     1019 |   948
     1020 |   941
     1021 |   943
     1022 |   946
     1023 |   949
     1024 |   949
     1025 |   959
     1026 |   940
     1027 |   952
     1028 |   964
     1029 |   940
     1030 |   961
          |     0
(31 rows)

-- (group by x::varchar)
select key::varchar(20),count(float_x)  from strom_test group by key::varchar(20) order by key;
 key | count 
-----+-------
 1   |   962
 10  |   960
 11  |   943
 12  |   952
 13  |   956
 14  |   945
 15  |   956
 16  |   952
 17  |   951
 18  |   955
 19  |   948
 2   |   952
 20  |   941
 21  |   943
 22  |   946
 23  |   949
 24  |   949
 25  |   959
 26  |   940
 27  |   952
 28  |   964
 29  |   940
 3   |   957
 30  |   961
 4   |   938
 5   |   945
 6   |   948
 7   |   947
 8   |   948
 9   |   954
     |     0
(31 rows)

-- (group by x::numeric)
select key::numeric,count(float_x)  from strom_test group by key::numeric order by key;
 key | count 
-----+-------
   1 |   962
   2 |   952
   3 |   957
   4 |   938
   5 |   945
   6 |   948
   7 |   947
   8 |   948
   9 |   954
  10 |   960
  11 |   943
  12 |   952
  13 |   956
  14 |   945
  15 |   956
  16 |   952
  17 |   951
  18 |   955
  19 |   948
  20 |   941
  21 |   943
  22 |   946
  23 |   949
  24 |   949
  25 |   959
  26 |   940
  27 |   952
  28 |   964
  29 |   940
  30 |   961
     |     0
(31 rows)

-- aggfunc( x + y + z )
select key, avg(float_x + float_y + float_z)  from strom_mix  group by key order by key;
 key |        avg        
-----+-------------------
   1 | 0.000211111484183
   2 |  -0.0121923785974
   3 |   0.0055784198712
   4 |   0.0586664406274
   5 |  -0.0219292143102
   6 |  -0.0152151418002
   7 |  0.00164136035847
   8 |  0.00193226280815
   9 |  0.00850789035607
  10 |   0.0108956559427
(10 rows)

